File: C++/November2022/901.online-stock-span.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=901 lang=cpp
 *
 * [901] Online Stock Span
 */

// @lc code=start
class StockSpanner
{
private:
    stack<pair<int, int>> stk;

public:
    StockSpanner()
    {
    }

    int next(int price)
    {
        int ans = 1;
        while (!stk.empty() && stk.top().first <= price)
        {
            ans += stk.top().second;
            stk.pop();
        }
        stk.push({price, ans});
        return ans;
    }
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */
// @lc code=end

File: Java/Misc/901.online-stock-span.java
package Misc;

import java.util.*;
/*
 * @lc app=leetcode id=901 lang=java
 *
 * [901] Online Stock Span
 */

// @lc code=start
class StockSpanner {
    private Stack<int[]> stk;

    public StockSpanner() {
        stk = new Stack<>();
    }

    public int next(int price) {
        int ans = 1;
        while (!stk.empty() && stk.peek()[1] <= price) {
            ans += stk.pop()[0];
        }
        stk.push(new int[] { ans, price });
        return ans;
    }
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner obj = new StockSpanner();
 * int param_1 = obj.next(price);
 */
// @lc code=end
